Add further changes for complex128 support

ibab opened the pull request:
I'm now back to working on extending the existing ops to support `complex128` where appropriate.
This addresses #1517.

This is not close to being complete yet, but I thought I should start a PR.

ibab committed:
Add further changes for complex128 support

gunan commented:
Can one of the admins verify this patch?

tensorflow-jenkins commented:
Can one of the admins verify this patch?

girving commented on tensorflow/core/kernels/cwise_op_complex.cc:
```diff
@@ -16,10 +16,17 @@ limitations under the License.
+                        .Device(DEVICE_CPU)
+                        .TypeConstraint<double>("T")
+                        .TypeConstraint<complex128>("Tout"),
+                        BinaryOp<CPUDevice, functor::make_complex<double>>);
```
`#define` a `REGISTER(R, C)` macro to avoid duplication.

girving commented on tensorflow/core/kernels/cwise_op_imag.cc:
```diff
@@ -16,10 +16,17 @@ limitations under the License.
+                        .Device(DEVICE_CPU)
+                        .TypeConstraint<complex128>("T")
+                        .TypeConstraint<double>("Tout"),
+                        UnaryOp<CPUDevice, functor::get_imag<complex128>>);
```
`REGISTER(R, C)`

girving commented on tensorflow/core/kernels/cwise_op_real.cc:
```diff
@@ -16,10 +16,17 @@ limitations under the License.
+                        .Device(DEVICE_CPU)
+                        .TypeConstraint<complex128>("T")
+                        .TypeConstraint<double>("Tout"),
+                        UnaryOp<CPUDevice, functor::get_real<complex128>>);
```
`REGISTER(R, C)`

girving commented on tensorflow/python/kernel_tests/cwise_ops_test.py:
```diff
@@ -65,19 +65,19 @@ def _compareCpu(self, x, np_func, tf_func):
 
-      if x.dtype == np.complex64 and tf_func in (
-          tf.sign, tf.sqrt, tf.rsqrt, tf.log):
+      if (x.dtype in (np.complex64, np.complex128) and \
```
Remove the \ at the end of the line, since you're already using parentheses.

girving replied:
Use x.dtype.is_complex

girving commented on tensorflow/python/kernel_tests/cwise_ops_test.py:
```diff
@@ -65,19 +65,19 @@ def _compareCpu(self, x, np_func, tf_func):
         return  # Return early
 
-      if x.dtype == np.float32 or x.dtype == np.complex64:
+      if x.dtype in (np.float32, np.complex64):
```
x.dtype.is_complex

girving replied:
Oops, no, cancel that.

girving commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -297,29 +297,91 @@ def complex(real, imag, name=None):
+    return gen_math_ops._complex(real, imag, Tout=Tout, name=name)
+
+
+def real(in_, name=None):
```
in_ -> value

girving commented on tensorflow/python/ops/math_ops.py:
```diff
@@ -297,29 +297,91 @@ def complex(real, imag, name=None):
+    return gen_math_ops.real(in_, Tout=in_.dtype.real_dtype, name=name)
+
+
+def imag(in_, name=None):
```
in_ -> value

girving commented on tensorflow/core/kernels/reduction_ops_sum.cc:
```diff
@@ -46,10 +46,22 @@ REGISTER_KERNEL_BUILDER(
+        .Device(DEVICE_GPU)
+        .TypeConstraint<complex128>("T")
+        .HostMemory("reduction_indices"),
+    ReductionOp<GPUDevice, complex64, Eigen::internal::SumReducer<complex64>>);
```
If you define a `REGISTER` macro, you'll probably automatically fix the bug you introduced by not defining one. :)

The second macro shouldn't mention complex64, I think.

ibab replied:
Oops, I didn't actually plan to put this into the PR, so I never took a second look at it. Thanks for spotting this.

girving commented:
Looks great!  Please smash once you fix the notes above to avoid the merge conflicted commit.

ibab commented:
Thanks for the suggestions!

vrv commented:
ping for update?

girving commented:
@vrv, @ibab: I was waiting for the smash.  Unfortunately now we have an overlapping PR here: https://github.com/tensorflow/tensorflow/pull/2263.

ibab commented:
Sorry, was extremely busy with my PhD for a few weeks.
@hunkim's overlapping PR is a bit strange.
Looks like it doesn't correctly handle the floating point precision when going from complex to float and vice versa.
(For example, taking the real part of a complex128 returns a single precision float)
Handling this correctly is most of the work in this PR.

I can address @girving's suggestions over the next few days if there's still interest in merging this.

girving commented:
Apologies, @ibab and @hunkim: I should have remembered the original PR when I started reviewing the new one.  Apologies also for forgetting about the precision issue.  @hunkim: Would you be okay with @ibab finishing up his PR since it handles the precision correctly?

hunkim commented:
@ibab Thanks for pointing it out. I was not aware of this PR. I think we should go for your PR first. 
@girving, absolutely.

ibab commented:
I rebased my changes on top of `master`, fixing the resulting merge conflicts and applying @girving's comments.
In order to get the tests to pass, I've converted a few more of the ops to `complex128`, but I tried to keep it minimal in order to not conflict with @hunkim's PR too much.

Also, I didn't commit the `ops.pbtxt`, as it seems that there are currently some unrelated ops on `master` for which the `ops.pbtxt` hasn't been committed yet.

hunkim commented:
@ibab if you want, feel free to grab my changes into your PR.

BTW, how was your PHD work? Good luck!

On Sat, May 14, 2016 at 4:30 PM, Igor Babuschkin <notifications@github.com>
wrote:

> I rebased my changes on top of master, fixing the resulting merge
> conflicts and applying @girving <https://github.com/girving>'s comments.
> In order to get the tests to pass, I've converted a few more of the ops to
> complex128, but I tried to keep it minimal in order to not conflict with
> @hunkim <https://github.com/hunkim>'s PR too much.
>
> Also, I didn't commit the ops.pbtxt, as it seems that there are currently
> some unrelated ops on master for which the ops.pbtxt hasn't been
> committed yet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1929#issuecomment-219252568>
>

ibab commented:
@hunkim:
> BTW, how was your PHD work? Good luck!

Thanks, it's going great.

I think we should try to keep the PRs as small as possible to make it easier to review them.
So we could continue with your PR, or make a new one after this.

hunkim commented:
Sounds good.

On Mon, May 16, 2016 at 2:58 AM, Igor Babuschkin <notifications@github.com>
wrote:

> @hunkim <https://github.com/hunkim>:
>
> BTW, how was your PHD work? Good luck!
>
> Thanks, it's going great.
>
> I think we should try to keep the PRs as small as possible to make it
> easier to review them.
> So we could continue with your PR, or make a new one after this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1929#issuecomment-219369787>
>

girving commented:
Looks good!  Jenkins, test this please.

girving commented:
Sweet, tests pass.  @ibab: Shall I go ahead and merge?

girving commented:
I went ahead and merged.  Further changes in further PRs.  Thank you for the contribution!

ibab commented:
Thanks!
@hunkim: Would you prefer if we continue development on #2263 or should I make a new PR?
Your PR contains many of the changes we would need for complete `complex128` support, but we would have to fix some merge conflicts.

hunkim commented:
I think it makes sense if you finish this complex128 change. Please go
ahead. Thanks.
On Mon, May 16, 2016 at 4:54 PM Igor Babuschkin <notifications@github.com>
wrote:

> Thanks!
> @hunkim <https://github.com/hunkim>: Would you prefer if we continue
> development on #2263 <https://github.com/tensorflow/tensorflow/pull/2263>
> or should I make a new PR?
> Your PR contains many of the changes we would need for complete complex128
> support, but we would have to fix some merge conflicts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/1929#issuecomment-219559733>
>
